### DISCUSSION TITLE
What is a `valid handle to an open stream object`?

### DISCUSSION THREAD

**Discussion by AvatarGanymede** [microsoft/msquic#2171] [Q&A] [answered]
I'm trying to implement a quic server with 4 connections to it. When sending data, I maintain a `ref_count` variable, before sending, `ref_count++`, when getting `SEND_COMPLETE` event, `ref_count--`. When `ref_count == 0` I free the send buffer.
So I try to pass a struct as the context of `StreamSend`:
```
typedef struct MANAGER {
	uint32_t ref_count;
	uint32_t Length;
	_Field_size_bytes_(Length)
	uint8_t* Buffer;
} MANAGER;
```
```
MsQuic->StreamSend(Stream, SendBuffer, 1, QUIC_SEND_FLAG_NONE | QUIC_SEND_FLAG_ALLOW_0_RTT, Manager))
```
In `StreamCallBack` funciton:
```
    MANAGER* ClientContext = (MANAGER*)(Event->SEND_COMPLETE.ClientContext);
    switch (Event->Type) {
    ...
    case QUIC_STREAM_EVENT_SEND_COMPLETE:
        ClientContext->ref_count --;
        if (ClientContext->ref_count == 0) {
            free(ClientContext);
        }
        break;
    default:
        break;
    }
```
The program failed, without any feedbacks.
In the [doc](https://github.com/microsoft/msquic/blob/4c593bf74f297545751b56fc7e8c1d93b975ed89/docs/api/StreamSend.md), it says

> `ClientSendContext`
> 
> The valid handle to an open stream object.

What is a "valid handle to an open stream object"?
I've read all the examples in this repo and found that the context of `StreamSend` is either `SendBuffer` or `nullptr`. Does this means that the "valid" SendContext should be SendBuffer?

**Comment by nibanks** [comment, marked as answer]
First off, it's a doc bug that it says `ClientSendContext` should be "The valid handle to an open stream object." It can be any data the application wants; and is simply a pass through, eventually being delivered to the app in the `QUIC_STREAM_EVENT_SEND_COMPLETE` events. So that wouldn't be the cause of your failure. It's likely some other variable you are passing into the function is bad. What exactly is failing? And with what error?

**Reply by AvatarGanymede** [reply]
OK, now it's confirmed that any data can be passed as SendContext, which is important for me. Next I will concentrate on debugging ( have tried for 3 hours though -_-). Thanks for your reply !
